File: course_android/DRecyclerView/app/src/main/java/better/drecyclerview/Adapter.java
package better.drecyclerview;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

/**
 * Created by better on 2017/1/18.
 */

public class Adapter extends RecyclerView.Adapter<Adapter.Holder> implements RVDragMoveHelper.DragDataListener {
    private List<String> ls;

    public Adapter(List<String> ls) {
        this.ls = ls;
    }

    @Override
    public Holder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new Holder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_txt, parent, false));
    }

    @Override
    public void onBindViewHolder(Holder holder, int position) {
        holder.txt.setText("" + position);

    }

    @Override
    public int getItemCount() {
        return 20;
    }

    @Override
    public List<?> getData() {
        return ls;
    }

    class Holder extends RecyclerView.ViewHolder {
        TextView txt;

        public Holder(View itemView) {
            super(itemView);
            txt = (TextView) itemView.findViewById(R.id.txt);
        }
    }
}

File: course_android/Rv/app/src/main/java/com/test/rv/d1/Adapter.kt
package com.test.rv.d1

import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import com.test.rv.BaseAdapter
import com.test.rv.BaseViewHolder
import com.test.rv.databinding.ItemDemo1Binding

class Adapter : BaseAdapter<Int, ItemDemo1Binding>() {
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): BaseViewHolder<ItemDemo1Binding> = BaseViewHolder(
        ItemDemo1Binding.inflate(LayoutInflater.from(parent.context), parent, false)
    ).also {
        Log.d("Demo1", "onCreateViewHolder() ${it.hashCode()}")
        val s =
            Thread.currentThread().stackTrace.joinToString("\n") { stackTrace -> stackTrace.className + ":" + stackTrace.methodName }
        Log.d("Demo1", s)
    }

    override fun onBindViewHolder(holder: BaseViewHolder<ItemDemo1Binding>, position: Int) {
        if (holder.itemView.tag !is Int) {
            holder.itemView.tag = position
        }
        Log.d(
            "Demo1",
            "Bind() $position holder:${holder.hashCode()} create from index: ${holder.itemView.tag}"
        )
        val s =
            Thread.currentThread().stackTrace.joinToString("\n") { stackTrace -> stackTrace.className + ":" + stackTrace.methodName }
        Log.d("Demo1", "$position: $s")
        holder.binding.title.text = position.toString()
    }
}